File: backend/pkg/api/agent/handlers.go
package agent

import (
	"backend/pkg/api/response"
	agentws "backend/pkg/api/websocket"
	"backend/pkg/core"
	coreAgent "backend/pkg/core/agent"
	"backend/pkg/core/chat"
	"backend/pkg/core/worker"
	"backend/pkg/database"
	"context"
	"encoding/json"
	"log"
	"strconv"
	"time"

	websocketLib "github.com/gofiber/contrib/websocket"
	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
)

// chatSvc is a lazy-initialized chat service
var chatSvc *chat.MessageService

func getChatService() *chat.MessageService {
	if chatSvc == nil {
		chatSvc = chat.NewMessageService(database.New())
	}
	return chatSvc
}

// AgentCopilot handles POST /agent
// @Summary Submit agent request
// @Description Submit a chat request to the AI agent for processing
// @Tags agent
// @Accept json
// @Produce json
// @Param request body object true "Chat request"
// @Success 200 {object} response.SuccessResponse{data=object{request_id=string,status=string}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /agent [post]
func AgentCopilot(c *fiber.Ctx) error {
	var req coreAgent.ChatRequest
	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}

	manager := coreAgent.GetAgentAsyncCopilotManager()
	requestID, err := manager.SubmitChatRequest(req)
	if err != nil {
		log.Printf("[Agent API] Failed to submit request: %v", err)
		return response.Error(c, err)
	}

	return response.Success(c, coreAgent.ChatRequestResponse{
		RequestID: requestID,
		Status:    "processing",
	})
}

// WebsocketHandler handles GET /websocket
func WebsocketHandler(con *websocketLib.Conn) {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	agentManager := coreAgent.GetAgentAsyncCopilotManager()
	
	// Track worker status subscription
	var workerStatusSubscribed bool
	var workerStatusCancel context.CancelFunc

	go func() {
		defer cancel()
		for {
			messageType, message, err := con.ReadMessage()
			if err != nil {
				break
			}
			if messageType == websocketLib.TextMessage {
				var msg struct {
					RequestID string `json:"requestId"`
					Action    string `json:"action"`
					Channel   string `json:"channel"`
				}
				if err := json.Unmarshal(message, &msg); err != nil {
					continue
				}

				// Handle channel subscriptions (e.g., worker-status)
				if msg.Channel == agentws.ChannelWorkerStatus {
					if msg.Action == "subscribe" && !workerStatusSubscribed {
						workerStatusSubscribed = true
						var wsCtx context.Context
						wsCtx, workerStatusCancel = context.WithCancel(ctx)
						go handleWorkerStatusStream(wsCtx, con)
						
						// Send acknowledgment
						ack := map[string]interface{}{
							"type":    "subscribed",
							"channel": agentws.ChannelWorkerStatus,
						}
						if data, err := json.Marshal(ack); err == nil {
							con.WriteMessage(websocketLib.TextMessage, data)
						}
					} else if msg.Action == "unsubscribe" && workerStatusSubscribed {
						workerStatusSubscribed = false
						if workerStatusCancel != nil {
							workerStatusCancel()
						}
					}
					continue
				}

				// Handle request-based subscriptions (agent streams)
				if msg.Action == "subscribe" && msg.RequestID != "" {
					agentws.HandleAgentStream(ctx, con, msg.RequestID, agentManager)
				}
			}
		}
	}()
	<-ctx.Done()
	
	// Cleanup worker status subscription
	if workerStatusCancel != nil {
		workerStatusCancel()
	}
}

// handleWorkerStatusStream streams worker status updates to a WebSocket connection
func handleWorkerStatusStream(ctx context.Context, con *websocketLib.Conn) {
	statusService := worker.GetStatusService()
	subscriber := statusService.Subscribe()
	defer statusService.Unsubscribe(subscriber)

	// Send initial status of all workers
	for name, status := range statusService.GetAllStatuses() {
		msg := formatWorkerStatusMessage(name, &status)
		if data, err := json.Marshal(msg); err == nil {
			if err := con.WriteMessage(websocketLib.TextMessage, data); err != nil {
				return
			}
		}
	}

	// Stream updates
	for {
		select {
		case update, ok := <-subscriber:
			if !ok {
				return
			}
			msg := formatWorkerStatusMessage(update.WorkerName, &update.Status)
			msg["timestamp"] = update.Timestamp
			if data, err := json.Marshal(msg); err == nil {
				if err := con.WriteMessage(websocketLib.TextMessage, data); err != nil {
					return
				}
			}
		case <-ctx.Done():
			return
		}
	}
}

// formatWorkerStatusMessage formats a worker status for WebSocket transmission
func formatWorkerStatusMessage(workerName string, status *worker.WorkerStatus) map[string]interface{} {
	msg := map[string]interface{}{
		"type":        "worker-status",
		"worker_name": workerName,
		"status": map[string]interface{}{
			"name":        status.Name,
			"state":       string(status.State),
			"progress":    status.Progress,
			"message":     status.Message,
			"error":       status.Error,
			"items_total": status.ItemsTotal,
			"items_done":  status.ItemsDone,
		},
	}

	if status.StartedAt != nil {
		msg["status"].(map[string]interface{})["started_at"] = status.StartedAt.Format(time.RFC3339)
	}
	if status.CompletedAt != nil {
		msg["status"].(map[string]interface{})["completed_at"] = status.CompletedAt.Format(time.RFC3339)
	}

	return msg
}

// GetConversationHistory handles GET /agent/conversations/:articleId
// @Summary Get conversation history
// @Description Get the chat history for an article
// @Tags agent
// @Accept json
// @Produce json
// @Param articleId path string true "Article ID"
// @Param limit query int false "Max messages to return" default(50)
// @Success 200 {object} response.SuccessResponse{data=object{messages=[]object,article_id=string,total=int}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /agent/conversations/{articleId} [get]
func GetConversationHistory(c *fiber.Ctx) error {
	articleID := c.Params("articleId")
	if articleID == "" {
		return response.Error(c, core.InvalidInputError("Article ID is required"))
	}

	log.Printf("[ConversationHistory] Fetching messages for article: %s", articleID)

	articleUUID, err := uuid.Parse(articleID)
	if err != nil {
		log.Printf("[ConversationHistory] Invalid article ID format: %s", articleID)
		return response.Error(c, core.InvalidInputError("Invalid article ID format"))
	}

	limit := 50
	if limitStr := c.Query("limit"); limitStr != "" {
		if parsedLimit, err := strconv.Atoi(limitStr); err == nil && parsedLimit > 0 {
			limit = parsedLimit
		}
	}

	messages, err := getChatService().GetConversationHistory(c.Context(), articleUUID, limit)
	if err != nil {
		log.Printf("[ConversationHistory] Failed to fetch messages: %v", err)
		return response.Error(c, err)
	}

	log.Printf("[ConversationHistory] Found %d messages for article %s", len(messages), articleID)

	return response.Success(c, fiber.Map{
		"messages":   messages,
		"article_id": articleID,
		"total":      len(messages),
	})
}

// ClearConversationHistory handles DELETE /agent/conversations/:articleId
// @Summary Clear conversation history
// @Description Clear all chat messages for an article
// @Tags agent
// @Accept json
// @Produce json
// @Param articleId path string true "Article ID"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /agent/conversations/{articleId} [delete]
func ClearConversationHistory(c *fiber.Ctx) error {
	articleID := c.Params("articleId")
	if articleID == "" {
		return response.Error(c, core.InvalidInputError("Article ID is required"))
	}

	log.Printf("[ConversationHistory] Clearing messages for article: %s", articleID)

	articleUUID, err := uuid.Parse(articleID)
	if err != nil {
		log.Printf("[ConversationHistory] Invalid article ID format: %s", articleID)
		return response.Error(c, core.InvalidInputError("Invalid article ID format"))
	}

	if err := getChatService().ClearConversationHistory(c.Context(), articleUUID); err != nil {
		log.Printf("[ConversationHistory] Failed to clear messages: %v", err)
		return response.Error(c, err)
	}

	log.Printf("[ConversationHistory] Successfully cleared messages for article %s", articleID)

	return response.Success(c, fiber.Map{"success": true})
}

// GetPendingArtifacts handles GET /agent/artifacts/:articleId/pending
// @Summary Get pending artifacts
// @Description Get all pending artifacts for an article
// @Tags agent
// @Accept json
// @Produce json
// @Param articleId path string true "Article ID"
// @Success 200 {object} response.SuccessResponse{data=object{artifacts=[]object}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /agent/artifacts/{articleId}/pending [get]
func GetPendingArtifacts(c *fiber.Ctx) error {
	articleIDStr := c.Params("articleId")
	articleID, err := uuid.Parse(articleIDStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid article ID"))
	}

	artifacts, err := getChatService().GetPendingArtifacts(c.Context(), articleID)
	if err != nil {
		return response.Error(c, err)
	}

	return response.Success(c, fiber.Map{
		"artifacts": artifacts,
	})
}

// AcceptArtifact handles POST /agent/artifacts/:messageId/accept
// @Summary Accept artifact
// @Description Accept a pending artifact
// @Tags agent
// @Accept json
// @Produce json
// @Param messageId path string true "Message ID"
// @Param request body object{feedback=string} false "Optional feedback"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /agent/artifacts/{messageId}/accept [post]
func AcceptArtifact(c *fiber.Ctx) error {
	messageIDStr := c.Params("messageId")
	messageID, err := uuid.Parse(messageIDStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid message ID"))
	}

	var req struct {
		Feedback string `json:"feedback"`
	}
	_ = c.BodyParser(&req)

	if err := getChatService().AcceptArtifact(c.Context(), messageID, req.Feedback); err != nil {
		return response.Error(c, err)
	}

	return response.Success(c, fiber.Map{"success": true})
}

// RejectArtifact handles POST /agent/artifacts/:messageId/reject
// @Summary Reject artifact
// @Description Reject a pending artifact
// @Tags agent
// @Accept json
// @Produce json
// @Param messageId path string true "Message ID"
// @Param request body object{feedback=string} false "Optional feedback"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /agent/artifacts/{messageId}/reject [post]
func RejectArtifact(c *fiber.Ctx) error {
	messageIDStr := c.Params("messageId")
	messageID, err := uuid.Parse(messageIDStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid message ID"))
	}

	var req struct {
		Feedback string `json:"feedback"`
	}
	_ = c.BodyParser(&req)

	if err := getChatService().RejectArtifact(c.Context(), messageID, req.Feedback); err != nil {
		return response.Error(c, err)
	}

	return response.Success(c, fiber.Map{"success": true})
}

File: backend/src/api/agent/handlers.rs
use axum::{
    Json,
    body::Bytes,
    extract::{Path, Query, State},
};
use uuid::Uuid;

use crate::{
    api::{auth::AuthenticatedAccount, request::JsonBody, response::SuccessResponse},
    error::AppError,
};

use super::{
    dto::{
        AgentToolListResponse, AgentToolResponse, ArtifactFeedbackRequest, ChatRequest,
        ChatRequestResponse, ConnectorListResponse, ConnectorRefreshResponse, ConnectorResponse,
        ConnectorUpdateRequest, ConnectorWriteRequest, ConversationHistoryResponse,
        ConversationQuery, ConversationTurnRequest, ConversationTurnResponse,
        PendingArtifactsResponse, SuccessFlagResponse,
    },
    state::AgentState,
};

type ApiResult<T> = Result<Json<SuccessResponse<T>>, AppError>;

#[utoipa::path(
    post,
    path = "/agent",
    request_body = ChatRequest,
    responses(
        (status = 200, body = SuccessResponse<ChatRequestResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "agent",
    operation_id = "submitAgentRequest"
)]
pub async fn submit_agent_request(
    _authenticated: AuthenticatedAccount,
    State(state): State<AgentState>,
    JsonBody(request): JsonBody<ChatRequest>,
) -> ApiResult<ChatRequestResponse> {
    if request.message.is_empty() {
        return Err(AppError::InvalidInput(
            "message is a required field".to_owned(),
        ));
    }
    if request.article_id.is_empty() {
        return Err(AppError::InvalidInput(
            "articleId is a required field".to_owned(),
        ));
    }
    let request_id = state.requests()?.submit(request).await?;
    Ok(Json(SuccessResponse::new(ChatRequestResponse {
        request_id,
        status: "processing".to_owned(),
    })))
}

#[utoipa::path(
    post,
    path = "/agent/conversation/turn",
    request_body = ConversationTurnRequest,
    responses(
        (status = 200, body = SuccessResponse<ConversationTurnResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "agent",
    operation_id = "submitConversationTurn"
)]
pub async fn submit_conversation_turn(
    _authenticated: AuthenticatedAccount,
    State(state): State<AgentState>,
    JsonBody(request): JsonBody<ConversationTurnRequest>,
) -> ApiResult<ConversationTurnResponse> {
    if request.article_id.is_empty() {
        return Err(AppError::InvalidInput(
            "articleId is a required field".to_owned(),
        ));
    }
    let turn = state
        .conversation()?
        .prepare_turn(&request.message, &request.audio_base64, &request.mime_type)
        .await?;
    let request_id = state
        .requests()?
        .submit(ChatRequest {
            message: turn.message,
            document_content: request.document_content,
            document_markdown: request.document_markdown,
            article_id: request.article_id,
            channel: turn.channel.clone(),
        })
        .await?;
    Ok(Json(SuccessResponse::new(ConversationTurnResponse {
        request_id,
        status: "processing".to_owned(),
        transcript: turn.transcript,
        channel: turn.channel,
    })))
}

#[utoipa::path(
    get,
    path = "/agent/tools",
    responses(
        (status = 200, body = SuccessResponse<AgentToolListResponse>),
        (status = 401, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "agent",
    operation_id = "listAgentTools"
)]
pub async fn list_agent_tools(
    _authenticated: AuthenticatedAccount,
    State(state): State<AgentState>,
) -> ApiResult<AgentToolListResponse> {
    let tools = state
        .registry()?
        .describe()
        .into_iter()
        .map(AgentToolResponse::from)
        .collect();
    Ok(Json(SuccessResponse::new(AgentToolListResponse { tools })))
}

#[utoipa::path(
    get,
    path = "/agent/connectors",
    responses(
        (status = 200, body = SuccessResponse<ConnectorListResponse>),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "agent",
    operation_id = "listMcpConnectors"
)]
pub async fn list_mcp_connectors(
    _authenticated: AuthenticatedAccount,
    State(state): State<AgentState>,
) -> ApiResult<ConnectorListResponse> {
    let connectors = state
        .connectors()?
        .list()
        .await?
        .into_iter()
        .map(ConnectorResponse::from)
        .collect();
    Ok(Json(SuccessResponse::new(ConnectorListResponse {
        connectors,
    })))
}

#[utoipa::path(
    post,
    path = "/agent/connectors",
    request_body = ConnectorWriteRequest,
    responses(
        (status = 200, body = SuccessResponse<ConnectorResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "agent",
    operation_id = "createMcpConnector"
)]
pub async fn create_mcp_connector(
    _authenticated: AuthenticatedAccount,
    State(state): State<AgentState>,
    JsonBody(request): JsonBody<ConnectorWriteRequest>,
) -> ApiResult<ConnectorResponse> {
    let connector = state.connectors()?.create(request.into()).await?;
    Ok(Json(SuccessResponse::new(connector.into())))
}

#[utoipa::path(
    patch,
    path = "/agent/connectors/{connectorId}",
    params(("connectorId" = String, Path)),
    request_body = ConnectorUpdateRequest,
    responses(
        (status = 200, body = SuccessResponse<ConnectorResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "agent",
    operation_id = "updateMcpConnector"
)]
pub async fn update_mcp_connector(
    _authenticated: AuthenticatedAccount,
    State(state): State<AgentState>,
    Path(connector_id): Path<String>,
    JsonBody(request): JsonBody<ConnectorUpdateRequest>,
) -> ApiResult<ConnectorResponse> {
    let id = parse_connector_id(&connector_id)?;
    let connector = state.connectors()?.update(id, request.into()).await?;
    Ok(Json(SuccessResponse::new(connector.into())))
}

#[utoipa::path(
    delete,
    path = "/agent/connectors/{connectorId}",
    params(("connectorId" = String, Path)),
    responses(
        (status = 200, body = SuccessResponse<SuccessFlagResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "agent",
    operation_id = "deleteMcpConnector"
)]
pub async fn delete_mcp_connector(
    _authenticated: AuthenticatedAccount,
    State(state): State<AgentState>,
    Path(connector_id): Path<String>,
) -> ApiResult<SuccessFlagResponse> {
    let id = parse_connector_id(&connector_id)?;
    state.connectors()?.delete(id).await?;
    Ok(Json(SuccessResponse::new(SuccessFlagResponse {
        success: true,
    })))
}

#[utoipa::path(
    post,
    path = "/agent/connectors/{connectorId}/refresh",
    params(("connectorId" = String, Path)),
    responses(
        (status = 200, body = SuccessResponse<ConnectorRefreshResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "agent",
    operation_id = "refreshMcpConnector"
)]
pub async fn refresh_mcp_connector(
    _authenticated: AuthenticatedAccount,
    State(state): State<AgentState>,
    Path(connector_id): Path<String>,
) -> ApiResult<ConnectorRefreshResponse> {
    let id = parse_connector_id(&connector_id)?;
    let refresh = state.connectors()?.refresh(id).await?;
    Ok(Json(SuccessResponse::new(ConnectorRefreshResponse {
        connector_id: refresh.connector_id.to_string(),
        tool_names: refresh.tool_names,
    })))
}

#[utoipa::path(
    get,
    path = "/agent/conversations/{articleId}",
    params(
        ("articleId" = String, Path),
        ConversationQuery
    ),
    responses(
        (status = 200, body = SuccessResponse<ConversationHistoryResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "agent",
    operation_id = "getConversationHistory"
)]
pub async fn get_conversation_history(
    _authenticated: AuthenticatedAccount,
    State(state): State<AgentState>,
    Path(article_id): Path<String>,
    Query(query): Query<ConversationQuery>,
) -> ApiResult<ConversationHistoryResponse> {
    if article_id.is_empty() {
        return Err(AppError::InvalidInput("Article ID is required".to_owned()));
    }
    let parsed_id = Uuid::parse_str(&article_id)
        .map_err(|_| AppError::InvalidInput("Invalid article ID format".to_owned()))?;
    let limit = query
        .limit
        .as_deref()
        .and_then(|value| value.parse::<i64>().ok())
        .filter(|value| *value > 0)
        .unwrap_or(50);
    let messages = state
        .chat()?
        .conversation_history(parsed_id, limit)
        .await?
        .into_iter()
        .map(Into::into)
        .collect::<Vec<_>>();
    let total = messages.len();
    Ok(Json(SuccessResponse::new(ConversationHistoryResponse {
        messages,
        article_id,
        total,
    })))
}

#[utoipa::path(
    delete,
    path = "/agent/conversations/{articleId}",
    params(("articleId" = String, Path)),
    responses(
        (status = 200, body = SuccessResponse<SuccessFlagResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "agent",
    operation_id = "clearConversationHistory"
)]
pub async fn clear_conversation_history(
    _authenticated: AuthenticatedAccount,
    State(state): State<AgentState>,
    Path(article_id): Path<String>,
) -> ApiResult<SuccessFlagResponse> {
    if article_id.is_empty() {
        return Err(AppError::InvalidInput("Article ID is required".to_owned()));
    }
    let article_id = Uuid::parse_str(&article_id)
        .map_err(|_| AppError::InvalidInput("Invalid article ID format".to_owned()))?;
    state.chat()?.clear_conversation_history(article_id).await?;
    Ok(Json(SuccessResponse::new(SuccessFlagResponse {
        success: true,
    })))
}

#[utoipa::path(
    get,
    path = "/agent/artifacts/{articleId}/pending",
    params(("articleId" = String, Path)),
    responses(
        (status = 200, body = SuccessResponse<PendingArtifactsResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "agent",
    operation_id = "getPendingArtifacts"
)]
pub async fn get_pending_artifacts(
    _authenticated: AuthenticatedAccount,
    State(state): State<AgentState>,
    Path(article_id): Path<String>,
) -> ApiResult<PendingArtifactsResponse> {
    let article_id = Uuid::parse_str(&article_id)
        .map_err(|_| AppError::InvalidInput("Invalid article ID".to_owned()))?;
    let artifacts = state
        .chat()?
        .pending_artifacts(article_id)
        .await?
        .into_iter()
        .map(Into::into)
        .collect();
    Ok(Json(SuccessResponse::new(PendingArtifactsResponse {
        artifacts,
    })))
}

#[utoipa::path(
    post,
    path = "/agent/artifacts/{messageId}/accept",
    params(("messageId" = String, Path)),
    request_body(content = ArtifactFeedbackRequest, content_type = "application/json"),
    responses(
        (status = 200, body = SuccessResponse<SuccessFlagResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "agent",
    operation_id = "acceptArtifact"
)]
pub async fn accept_artifact(
    _authenticated: AuthenticatedAccount,
    State(state): State<AgentState>,
    Path(message_id): Path<String>,
    body: Bytes,
) -> ApiResult<SuccessFlagResponse> {
    let message_id = parse_message_id(&message_id)?;
    let request = optional_feedback(&body)?;
    state
        .chat()?
        .accept_artifact(message_id, request.feedback)
        .await?;
    Ok(Json(SuccessResponse::new(SuccessFlagResponse {
        success: true,
    })))
}

#[utoipa::path(
    post,
    path = "/agent/artifacts/{messageId}/reject",
    params(("messageId" = String, Path)),
    request_body(content = ArtifactFeedbackRequest, content_type = "application/json"),
    responses(
        (status = 200, body = SuccessResponse<SuccessFlagResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "agent",
    operation_id = "rejectArtifact"
)]
pub async fn reject_artifact(
    _authenticated: AuthenticatedAccount,
    State(state): State<AgentState>,
    Path(message_id): Path<String>,
    body: Bytes,
) -> ApiResult<SuccessFlagResponse> {
    let message_id = parse_message_id(&message_id)?;
    let request = optional_feedback(&body)?;
    state
        .chat()?
        .reject_artifact(message_id, request.feedback)
        .await?;
    Ok(Json(SuccessResponse::new(SuccessFlagResponse {
        success: true,
    })))
}

fn parse_connector_id(connector_id: &str) -> Result<Uuid, AppError> {
    Uuid::parse_str(connector_id)
        .map_err(|_| AppError::InvalidInput("Invalid connector ID".to_owned()))
}

fn parse_message_id(message_id: &str) -> Result<Uuid, AppError> {
    Uuid::parse_str(message_id).map_err(|_| AppError::InvalidInput("Invalid message ID".to_owned()))
}

fn optional_feedback(body: &[u8]) -> Result<ArtifactFeedbackRequest, AppError> {
    if body.is_empty() {
        return Ok(ArtifactFeedbackRequest::default());
    }
    serde_json::from_slice(body)
        .map_err(|_| AppError::InvalidInput("Invalid request body".to_owned()))
}
